app/inpututils.cpp

```diff
@@ -1302,52 +1312,111 @@ QgsQuickMapSettings *InputUtils::setupMapSettings( QgsProject *project, QgsQuick
+
+  // We currently support only point geometries
+  if ( targetFeature.layer()->geometryType() != QgsWkbTypes::PointGeometry )
+    return extent;
+
+  if ( !mapPositioner->positionKit() || !mapPositioner->mapSettings() )
+    return extent;
+
+  //
+  // In order to compute stakeout extent, we first compute distance to target feature and
+  // based on that we update the extent like this:
+  //   - if distance is > 10m, use 5m scale and center to GPS point
+  //   - if distance is 3-10m, use 2m scale and center to GPS point
+  //   - if distance is 1-3m, use 1m scale and center to GPS point
+  //   - if distance is < 1m, use 50cm scale and center to target point (so that canvas does not move all the time)
+  //
+
+  // transform gps point to map's CRS and screen's X/Y pixels
+  QgsPoint gpsPointRaw = mapPositioner->positionKit()->positionCoordinate();
+  QgsPointXY gpsPointInMapCRS = transformPoint(
+                                  coordinateReferenceSystemFromEpsgId( 4326 ),
+                                  mapSettings->destinationCrs(),
+                                  mapSettings->transformContext(),
+                                  gpsPointRaw
+                                );
+
+  if ( gpsPointInMapCRS.isEmpty() )
   {
-    try
-    {
-      gpsPosition.transform( ct1 );
-    }
-    catch ( QgsCsException &e )
-    {
-      Q_UNUSED( e )
-      return QgsRectangle();
-    }
+    // unsuccessful transform
+    return extent;
   }
 
-  // Transform target point to map CRS
-  QgsPoint targetPoint( extractPointFromFeature( targetFeature ) );
-  QgsCoordinateTransform ct2( targetFeature.layer()->crs(), mapSettings->destinationCrs(), mapSettings->transformContext() );
-  if ( !ct2.isShortCircuited() )
+  QgsPointXY gpsPointInCanvasXY = mapPositioner->screenPosition();
+
+  qreal distance = distanceBetweenGpsAndFeature( gpsPointRaw, targetFeature, mapSettings );
+  qreal scale = 205; // ~ 5m scale
+  qreal panelOffset = 0; // (px) used as an offset in order to center point in visible extent (we center to gpsPoint/target point + this offset)
+
+  if ( mapExtentOffset > 0 )
   {
-    try
-    {
-      targetPoint.transform( ct2 );
-    }
-    catch ( QgsCsException &e )
+    panelOffset = mapExtentOffset / 2;
+  }
+
+  if ( distance > 10 )
+  {
+    QgsPointXY centerInCanvasXY( gpsPointInCanvasXY.x(), gpsPointInCanvasXY.y() + panelOffset );
+    QgsPointXY center = mapSettings->screenToCoordinate( centerInCanvasXY.toQPointF() );
+
+    extent = mapSettings->mapSettings().computeExtentForScale( center, scale );
+  }
+  else if ( distance <= 10 && distance > 3 )
+  {
+    scale = 105; // ~ 2m scale
+
+    QgsPointXY centerInCanvasXY( gpsPointInCanvasXY.x(), gpsPointInCanvasXY.y() + panelOffset );
+    QgsPointXY center = mapSettings->screenToCoordinate( centerInCanvasXY.toQPointF() );
+
+    extent = mapSettings->mapSettings().computeExtentForScale( center, scale );
+  }
+  else if ( distance <= 3 && distance > 1 )
```

Comment:
up to here it seems like action is the same only scale differs, maybe better to have distance2scale mapping and then just do the action (less repetitive)?

Reply:
Addressed